README.md
# choroplethrZip

v1.5.0: [![Build Status](https://travis-ci.org/arilamstein/choroplethrZip.svg?branch=v1.4.0)](https://travis-ci.org/arilamstein/choroplethrZip/branches)
Master: [![Build Status](https://travis-ci.org/arilamstein/choroplethrZip.svg)](https://travis-ci.org/arilamstein/choroplethrZip)

R package with shapefile, metadata and visualization functions for US Zip Code Tabulated Areas (ZCTAs).

## Training & Development

My free course [Learn to Map Census Data in R](http://www.arilamstein.com/free-course) can teach you how to use this package. I blog about this package's development [here](http://www.arilamstein.com/blog).

## Installation

Due to this package's size (~60MB) it is hosted on github, not CRAN. To install the latest stable release type the following from an R console:

```
# install.packages("devtools")
library(devtools)
install_github('arilamstein/choroplethrZip@v1.5.0')
```

## Usage

This package ships with two vignettes which explain how to use the package. For convenience I have also published them online:

1. [Creating ZIP Code Choropleths with *choroplethrZip*](http://rpubs.com/arilamstein/zip-vignette-1)
1. [Creating ZIP Code Choropleths of US Census Data with *choroplethrZip*](http://rpubs.com/arilamstein/zip-vignette-2)
2. [Zip Code Demographics with choroplethrZip v1.3.0](http://rpubs.com/arilamstein/zip-vignette-3)

In short: type `?zip_choropleth` and `?zip_choropleth_acs` after installing the package to get started.


NEWS
choroplethrZip 1.5.0
-------------------------------
-Fix bugs with get_zip_demographics

choroplethrZip 1.4.0
-------------------------------
-function ?zip_choropleth now has a parameter reference_map which defaults to false. If set to true it will add a reference map from google maps.

choroplethrZip 1.3.0
-------------------------------
-add function ?get_zip_demographics which returns selected demographic statistics about each Zip Code Tabulated Area (ZCTA) the US. 
-add data.frame df_zip_demographics which contains selected demographic statistics about each ZCTA from the US Census's 2013 5-year ACS.

choroplethrZip 1.2.2
-------------------------------
-update vignettes for national maps
-add in some missing data (e.g. news for v1.2.1)

choroplethrZip 1.2.1
-------------------------------
-don't run examples because they take too long for travis

choroplethrZip 1.2.0
-------------------------------
-zip_choropleth now supports nationwide choropleths. Note that nationwide choropleths can take a 
few minutes to render, so it is still recommended to use one of the various zooms.

choroplethrZip 1.1.0
-------------------------------
-add zip_choropleth function
-add zip_choropleth_acs function


tests/test-all.R
#library(testthat)

#test_package("choroplethrZip")


inst/tests/test-zip_choropleth.R
context("test zip_choropleth")

test_that("state_zoom returns ggplot", {
  data(df_pop_zip, package="choroplethrZip")
  expect_is(zip_choropleth(df_pop_zip, state_zoom=c("new york")), "ggplot")
})

test_that("setting title returns ggplot", {
  data(df_pop_zip, package="choroplethrZip")
  expect_is(zip_choropleth(df_pop_zip, state_zoom=c("new york"), title="test title"), "ggplot")
})

test_that("setting legend returns ggplot", {
  data(df_pop_zip, package="choroplethrZip")
  expect_is(zip_choropleth(df_pop_zip, state_zoom=c("new york"), legend="test legend"), "ggplot")
})

test_that("continuous scale returns ggplot", {
  data(df_pop_zip, package="choroplethrZip")
  expect_is(zip_choropleth(df_pop_zip, state_zoom=c("new york"), num_colors=1), "ggplot")
})

test_that("county_zoom returns ggplot", {
  data(df_pop_zip, package="choroplethrZip")
  nyc_fips = c(36005, 36047, 36061, 36081, 36085)
  expect_is(zip_choropleth(df_pop_zip,
                           county_zoom=nyc_fips,
                           title="2012 New York City Zip Population Estimates",
                           legend="Population"), "ggplot")
})

test_that("zip_zoom returns ggplot", {
  data(df_pop_zip, package="choroplethrZip")
  manhattan_les = c("10002", "10003", "10009")
  manhattan_ues = c("10021", "10028", "10044", "10128")
  zip_choropleth(df_pop_zip,
                 zip_zoom=c(manhattan_les, manhattan_ues),
                 title="2012 Lower and Upper East Side Zip Population Estimates",
                 legend="Population")
})

test_that("msa_zoom returns ggplot", {
  data(df_pop_zip, package="choroplethrZip")
  expect_is(zip_choropleth(df_pop_zip,
                 msa_zoom="New York-Newark-Jersey City, NY-NJ-PA",
                 title="2012 NY-Newark-Jersey City MSA\nZip Population Estimates",
                 legend="Population"), "ggplot")
})

# this test is important, but I think that it causes travis to fail because it can take a 
# few minutes to run
#test_that("full county zoom returns ggplot", {
#  data(df_pop_zip, package="choroplethrZip")
#  expect_is(zip_choropleth(df_pop_zip,
#                           title="2012 USA\nZip Population Estimates",
#                           legend="Population"), "ggplot")
#})

inst/tests/test-zip_choropleth_acs.R
context("test zip_choropleth_acs")

test_that("zip_choropleth_acs returns a ggplot2 with state_zoom set", {
  expect_is(
    suppressWarnings(zip_choropleth_acs("B19301", state_zoom="new york")),
    "ggplot")
})

test_that("zip_choropleth_acs returns a ggplot2 with county_zoom set", {
  nyc_fips = c(36005, 36047, 36061, 36081, 36085)
  expect_is(
    suppressWarnings(zip_choropleth_acs("B19301", county_zoom=nyc_fips)),
    "ggplot")
})

test_that("zip_choropleth_acs returns a ggplot2 with msa_zoom set", {
  expect_is(
    suppressWarnings(zip_choropleth_acs("B19301", num_colors=1, msa_zoom="New York-Newark-Jersey City, NY-NJ-PA")),
    "ggplot")
})

test_that("zip_choropleth_acs returns a ggplot2 with zip_zoom set", {
  # compare Manhattan's Lower East Side and Upper East Side
  manhattan_les = c("10002", "10003", "10009")
  manhattan_ues = c("10021", "10028", "10044", "10128")
  expect_is(
    suppressWarnings(zip_choropleth_acs("B19301", num_colors=1, zip_zoom=c(manhattan_les, manhattan_ues))),
    "ggplot")
})


R/acs.R
if (base::getRversion() >= "2.15.1") {
  utils::globalVariables(c("zip.regions"))
}

#' Create a US ZIP choropleth from ACS data
#' 
#' Creates a US ZIP choropleth using the US Census' American Community Survey (ACS) data.  
#' Requires the acs package to be installed, and a Census API Key to be set with 
#' the acs's api.key.install function.  Census API keys can be obtained at http://www.census.gov/developers/tos/key_request.html.
#'
#' @param tableId The id of an ACS table
#' @param endyear The end year of the survey to use.  See acs.fetch (?acs.fetch) and http://1.usa.gov/1geFSSj for details.
#' @param span The span of time to use.  See acs.fetch and http://1.usa.gov/1geFSSj for details.
#' on the same longitude and latitude map to scale. This variable is only checked when the "states" variable is equal to all 50 states.
#' @param num_colors The number of colors on the map. A value of 1 
#' will use a continuous scale. A value in [2, 9] will use that many colors. 
#' @param zip_zoom An optional vector of zip codes to zoom in on. Elements of this vector must exactly 
#' match the names of zips as they appear in the "region" column of ?zip.regions.
#' @param county_zoom An optional vector of county FIPS codes to zoom in on. Elements of this 
#' vector must exactly match the names of zips as they appear in the "county.fips.numeric" column 
#' of ?zip.regions.
#' @param state_zoom An optional vector of State names to zoom in on. Elements of this 
#' vector must exactly match the names of the state names as they appear in the "state.name" column 
#' of ?zip.regions.
#' @param msa_zoom An optional vector of MSA (Metroplitan/Micropolitan Statistical Area) names to zoom in on. Elements of this 
#' vector must exactly match the names of the state names as they appear in the "cbsa.title" column 
#' of ?zip.regions.
#' @return A choropleth.
#' 
#' @keywords choropleth, acs
#' 
#' @seealso \code{api.key.install} in the acs package which sets an Census API key for the acs library
#' @seealso http://factfinder2.census.gov/faces/help/jsf/pages/metadata.xhtml?lang=en&type=survey&id=survey.en.ACS_ACS 
#' which contains a list of all ACS surveys.
#' @references Uses the acs package created by Ezra Haber Glenn.
#' @export
#' @examples
#' \dontrun{
#' # Median income of all ZCTAs in New York State
#' zip_choropleth_acs("B19301", state_zoom="new york")
#' 
#' # zoom in on all ZCTAs in the 5 counties (boroughs) of New York City
#' nyc_fips = c(36005, 36047, 36061, 36081, 36085)
#' zip_choropleth_acs("B19301", county_zoom=nyc_fips)
#' 
#' # compare Manhattan's Lower East Side and Upper East Side
#' manhattan_les = c("10002", "10003", "10009")
#' manhattan_ues = c("10021", "10028", "10044", "10128")
#' zip_choropleth_acs("B19301", num_colors=1, zip_zoom=c(manhattan_les, manhattan_ues))
#' 
#' # show all ZCTAs in the New York Metropolitan Statistical Area (MSA)
#' zip_choropleth_acs("B19301", msa_zoom="New York-Newark-Jersey City, NY-NJ-PA")
#' }
#' @importFrom choroplethr get_acs_data
zip_choropleth_acs = function(tableId, endyear=2011, span=5, num_colors=7, state_zoom=NULL, county_zoom=NULL, msa_zoom=NULL, zip_zoom=NULL)
{
  acs.data = get_acs_data(tableId, "zip", endyear, span)
  
  # remove values that are not on the map to prevent the warning
  data(zip.regions, package="choroplethrZip", envir=environment())
  df = acs.data[[1]]
  df = df[df$region %in% unique(zip.regions$region), ]
  
  zip_choropleth(df, acs.data[['title']], "", num_colors,  state_zoom=state_zoom, county_zoom=county_zoom, msa_zoom=msa_zoom, zip_zoom=zip_zoom)
}

R/zip.R
#' Create a ZIP-level choropleth
#' 
#' @export
#' @importFrom R6 R6Class
ZipChoropleth = R6Class("ZipChoropleth",
  inherit = choroplethr:::Choropleth,
  public = list(
    
    # initialize with a world map
    initialize = function(user.df)
    {
      data(zcta.map, envir=environment())
      super$initialize(zcta.map, user.df)
    }
  )
)

#' Create a choropleth of USA Zip Code Tabulated Areas (ZCTAs) with sensible defaults.
#' 
#' The map used is zcta.map.  See zcta.regions for a list of valid regions.
#' 
#' @param df A data.frame with a column named "region" and a column named "value".  Elements in 
#' the "region" column must exactly match how regions are named in the "region" column in zcta.map.
#' @param title An optional title for the map.  
#' @param legend An optional name for the legend.  
#' @param buckets The number of equally sized buckets to places the values in.  A value of 1 
#' will use a continuous scale, and a value in [2, 9] will use that many buckets. 
#' @param zoom An optional vector of ZIPs to zoom in on. Elements of this vector must exactly 
#' match the names of ZIPs as they appear in the "region" column of ?zcta.regions
#' @export
zip_choropleth = function(df, title="", legend="", buckets=7, zoom=NULL)
{
  c = ZipChoropleth$new(df)
  c$title  = title
  c$legend = legend
  c$set_buckets(buckets)
  c$set_zoom(zoom)
  c$render()
}

R/data.R
#' Map of US Zip Code Tabulated Areas (ZCTAs)
#' 
#' The map
#' was modified using QGIS in order to remove Puerto Rico, islands off of Alaska that
#' crossed the antimeridian and American Samoa. A column named "region" was added for convenience which
#' is a duplicate of the column named ZCTA5CE10.
#'
#' @docType data
#' @name zip.map
#' @usage data(zip.map)
#' @references Taken from the US Census 2013 Cartographic Boundary Shapefiles - ZIP Code Tabulation Areas (ZCTAs): 
#' https://www.census.gov/geo/maps-data/data/cbf/cbf_zcta.html. 
NULL

#' Metadata about US Zip Code Tabulated Areas (ZCTAs)
#' 
#' The metadata is: State name, County name, County FIPS code, Core Based Statistical
#' Area (CBSA) id, CBSA title and 
#' Metropolitan/Micropolitan Statistical Area (MSA) name.
#' 
#' Note that this data.frame contains duplicates, because some ZCTAs span multiple Counties and States.
#' 
#' @seealso ?zip.map
#'  
#' @docType data
#' @name zip.regions
#' @usage data(zip.regions)
#' @references Data comes from the US Census Bureau's 2010 ZCTA to County Relationship File 
#' (https://www.census.gov/geo/maps-data/data/zcta_rel_download.html), 2010 ZCTA to Metropolitan and Micropolitan Statistical 
#' Areas Relationship File and the Core Based Statistical Areas (CBSAs) and Combined Statistical Areas (CSAs) Metropolitan 
#' and Micropolitan Delineation files (http://www.census.gov/population/metro/data/def.html).
NULL

#' A data.frame containing population estimates for US Zip Code Tabulated Areas (ZCTAs) in 2012.   
#' 
#' ZCTAs are intended to be roughly analogous to postal ZIP codes.
#'
#' @name df_pop_zip
#' @docType data
#' @references Taken from the US American Community Survey (ACS) 5 year estimates.
#' ZCTAs, and their realationship to ZIP codes, are explained here \url{https://www.census.gov/geo/reference/zctas.html}.
#' @keywords data
#' @usage data(df_pop_zip)
NULL

#' A data.frame containing demographic statistics for each Zip Code Tabulated Area (ZCTA).
#'
#' @name df_zip_demographics
#' @docType data
#' @references Data comes from the 2013 5-year American Community Survey (ACS). Data generated by
#' ?get_zip_demographics. 
#' @keywords data
#' @usage data(df_zip_demographics)
NULL


R/get_zip_demographics.R
#' Get a handful of demographic variables from the US Census Bureau as a data.frame.
#' 
#' The data comes from the American Community Survey (ACS). The geographic unit is 
#' Zip Code Tabulated Areas (ZCTAs). The variables are: total population, percent White 
#' not Hispanic, Percent Black or African American not Hispanic, percent Asian not Hispanic,
#' percent Hispanic all races, per-capita income, median rent and median age.
#' @param endyear The end year for the survey
#' @param span The span of the survey
#' @references The choroplethr guide to Census data: http://cran.r-project.org/web/packages/choroplethr/vignettes/e-mapping-us-census-data.html
#' @references A list of all ACS Surveys: http://factfinder.census.gov/faces/affhelp/jsf/pages/metadata.xhtml?lang=en&type=survey&id=survey.en.ACS_ACS
#' @importFrom acs geo.make acs.fetch geography estimate
#' @importFrom choroplethr get_acs_data
#' @export
get_zip_demographics = function(endyear=2013, span=5)
{
  zip_geo   = acs::geo.make(zip.code = "*")
  race.data = acs::acs.fetch(geography    = zip_geo, 
                             table.number = "B03002", 
                             col.names    = "pretty", 
                             endyear      = endyear, 
                             span         = span)
  
  # convert to a data.frame 
  df_race = data.frame(region                   = as.character(acs::geography(race.data)$zipcodetabulationarea), 
                       total_population         = as.numeric(acs::estimate(race.data[,1])),
                       white_alone_not_hispanic = as.numeric(acs::estimate(race.data[,3])),
                       black_alone_not_hispanic = as.numeric(acs::estimate(race.data[,4])),
                       asian_alone_not_hispanic = as.numeric(acs::estimate(race.data[,6])),
                       hispanic_all_races       = as.numeric(acs::estimate(race.data[,12])))
  
  df_race$region = as.character(df_race$region) # no idea why, but it's a factor before this line
  
  df_race$percent_white    = round(df_race$white_alone_not_hispanic / df_race$total_population * 100)
  df_race$percent_black    = round(df_race$black_alone_not_hispanic / df_race$total_population * 100)
  df_race$percent_asian    = round(df_race$asian_alone_not_hispanic / df_race$total_population * 100)
  df_race$percent_hispanic = round(df_race$hispanic_all_races       / df_race$total_population * 100)
  
  df_race = df_race[, c("region", "total_population", "percent_white", "percent_black", "percent_asian", "percent_hispanic")]
  
  # per capita income 
  df_income = choroplethr::get_acs_data("B19301", "zip", endyear=endyear, span=span)[[1]]  
  colnames(df_income)[[2]] = "per_capita_income"
  
  # median rent
  df_rent = get_acs_data("B25058", "zip", endyear=endyear, span=span)[[1]]  
  colnames(df_rent)[[2]] = "median_rent"
  
  # median age
  # can't do get_acs_data here because there seems to be a bug with how column_idx is treated right now
  #df_age = get_acs_data("B01002", "zip", endyear=endyear, span=span, column_idx=1)[[1]]  
  age    = acs::acs.fetch(geography=zip_geo, table.number = "B01002", col.names = "pretty", endyear = endyear, span = span)
  df_age = choroplethr:::convert_acs_obj_to_df("zip", age, 1, FALSE) 
  colnames(df_age)[[2]] = "median_age"
  
  df_demographics = merge(df_race        , df_income, all.x=TRUE)
  df_demographics = merge(df_demographics, df_rent  , all.x=TRUE)  
  df_demographics = merge(df_demographics, df_age   , all.x=TRUE)
  
  # remove the regions (such as zips in Puerto Rico) that are not on my map.
  data(zip.regions, package="choroplethrZip", envir=environment())
  df_demograhpics = df_demographics[df_demographics$region %in% zip.regions$region, ]
  
  df_demographics
}

R/zip_choropleth.R
#' An R6 object for creating ZIP-level choropleths.
#' @export
#' @importFrom dplyr left_join
#' @importFrom R6 R6Class
#' @importFrom choroplethr Choropleth
#' @importFrom grid grobTree
ZipChoropleth = R6Class("ZipChoropleth",
  inherit = Choropleth,
  public = list(
    
    # initialize with a ZIP map
    initialize = function(user.df)
    {
      data(zip.map, package="choroplethrZip", envir=environment())
      super$initialize(zip.map, user.df)
      
      if (private$has_invalid_regions)
      {
        warning("Please see ?zip.regions for a list of mappable regions")
      }
    },
    
    # All zooms, at the end of the day, are zip zooms. But often times it is more natural
    # for users to specify the zoom in other geographical units
    # This function name is a bit of a hack - it seems like I cannot override the parent set_zoom directly
    # because this function has a different number of parameters than that function, and the extra parameters
    # seeming just disappear
    set_zoom_zip = function(state_zoom, county_zoom, msa_zoom, zip_zoom)
    {
      # user can zoom by at most one of these options
      num_zooms_selected = sum(!is.null(c(state_zoom, county_zoom, msa_zoom, zip_zoom)))
      if (num_zooms_selected > 1) {
        stop("You can only zoom in by one of zip_zoom, county_zoom, state_zoom or msa_zoom")
      }
      
      data(zip.regions, package="choroplethrZip", envir=environment())
      
      # if the zip_zoom field is selected, just do default behavior
      if (!is.null(zip_zoom)) {
        super$set_zoom(zip_zoom)
      # if county_zoom field is selected, extract zips from counties  
      } else if (!is.null(county_zoom)) {
          stopifnot(all(county_zoom %in% unique(zip.regions$county.fips.numeric)))
          zips = zip.regions[zip.regions$county.fips.numeric %in% county_zoom, "region"]
          super$set_zoom(zips)        
      } else if (!is.null(state_zoom)) {
        stopifnot(all(state_zoom %in% unique(zip.regions$state.name)))
        zips = zip.regions[zip.regions$state.name %in% state_zoom, "region"]
        super$set_zoom(zips)        
      } else if (!is.null(msa_zoom)) {
        stopifnot(all(msa_zoom %in% unique(zip.regions$cbsa.title)))
        zips = zip.regions[zip.regions$cbsa.title %in% msa_zoom, "region"]
        super$set_zoom(zips)                
      }
    },
    
    render_nationwide = function()
    {
      data(zip.regions, package="choroplethrZip", envir=environment())

      self$prepare_map()
      
      # first render the continental us
      continental_zips   = zip.regions[!zip.regions$state.name %in% c("alaska", "hawaii"), "region"]
      continental.df     = self$choropleth.df[self$choropleth.df$region %in% continental_zips, ]
      continental.ggplot = self$render_helper(continental.df, self$scale_name, self$theme_clean()) + ggtitle(self$title)
      
      ret = continental.ggplot
      
      # render ak and add as inset
      ak_zips       = zip.regions[zip.regions$state.name == "alaska", "region"]
      alaska.df     = self$choropleth.df[self$choropleth.df$region %in% ak_zips, ]
      alaska.ggplot = self$render_helper(alaska.df, "", self$theme_inset())
      alaska.grob   = ggplotGrob(alaska.ggplot)
      ret           = ret + annotation_custom(grobTree(alaska.grob), xmin=-125, xmax=-110, ymin=22.5, ymax=30)

      # render hi and add as inset
      hi_zips       = zip.regions[zip.regions$state.name == "hawaii", "region"]
      hawaii.df     = self$choropleth.df[self$choropleth.df$region %in% hi_zips, ]
      hawaii.ggplot = self$render_helper(hawaii.df, "", self$theme_inset())
      hawaii.grob   = ggplotGrob(hawaii.ggplot)
      ret           = ret + annotation_custom(grobTree(hawaii.grob), xmin=-107.5, xmax=-102.5, ymin=25, ymax=27.5)

      ret +
        ggtitle(self$title)
    },
    
    render_helper = function(choropleth.df, scale_name, theme)
    {
      # maps with numeric values are mapped with a continuous scale
      if (is.numeric(choropleth.df$value))
      {
        ggplot(choropleth.df, aes(long, lat, group = group)) +
          geom_polygon(aes(fill = value), size = 0) + 
          self$get_scale() + 
          theme;
      } else { # assume character or factor
        stopifnot(length(unique(na.omit(choropleth.df$value))) <= 9) # brewer scale only goes up to 9
        
        ggplot(choropleth.df, aes(long, lat, group = group)) +
          geom_polygon(aes(fill = value), size = 0) + 
          self$get_scale() + 
          theme;
      }
    }
  )
)

#' Create a choropleth of US Zip Codes
#' 
#' Note that what is rendered are 2010 US Census Zip Code Tabulated Areas (ZCTAs), not USPS Zip Codes. The 
#' map used is zip.map in the choroplethrZip package. See zip.regions for
#' an object which can help you coerce your regions into the required format.
#' 
#' @param df A data.frame with a column named "region" and a column named "value".  Elements in 
#' the "region" column must exactly match how regions are named in the "region" column in ?zip.map.
#' @param title An optional title for the map.  
#' @param legend An optional name for the legend.  
#' @param num_colors The number of colors on the map. A value of 1 
#' will use a continuous scale. A value in [2, 9] will use that many colors. 
#' @param zip_zoom An optional vector of zip codes to zoom in on. Elements of this vector must exactly 
#' match the names of zips as they appear in the "region" column of ?zip.regions.
#' @param county_zoom An optional vector of county FIPS codes to zoom in on. Elements of this 
#' vector must exactly match the names of zips as they appear in the "county.fips.numeric" column 
#' of ?zip.regions.
#' @param state_zoom An optional vector of State names to zoom in on. Elements of this 
#' vector must exactly match the names of the state names as they appear in the "state.name" column 
#' of ?zip.regions.
#' @param msa_zoom An optional vector of MSA (Metroplitan/Micropolitan Statistical Area) names to zoom in on. Elements of this 
#' vector must exactly match the names of the state names as they appear in the "cbsa.title" column 
#' of ?zip.regions.
#' @param reference_map If true, render the choropleth over a reference map from Google Maps.
#' @note Nationwide zip choropleths can take a few minutes to render. 
#' It is much faster to view a subset of the country by selecting a zoom. 
#' @examples
#' \dontrun{
#' 
#' library(choroplethrZip)
#' data(df_pop_zip)
#'
#' # zooming on a state
#' zip_choropleth(df_pop_zip, 
#'                state_zoom = "new york", 
#'                title      = "2012 New York State ZCTA Population Estimates", 
#'                legend     = "Population")
#' 
#' # adding a reference map
#' zip_choropleth(df_pop_zip, 
#'                state_zoom    = "new york", 
#'                title         = "2012 New York State ZCTA Population Estimates", 
#'                legend        = "Population",
#'                reference_map = TRUE)
#'
#' # viewing on a set of counties
#' # note we use numeric county FIPS codes
#' nyc_fips = c(36005, 36047, 36061, 36081, 36085)
#' zip_choropleth(df_pop_zip, 
#'                county_zoom = nyc_fips, 
#'                title       = "2012 New York City ZCTA Population Estimates", 
#'                legend      = "Population")
#'
#' # zooming in on a few ZIPs
#' manhattan_les = c("10002", "10003", "10009")
#' manhattan_ues = c("10021", "10028", "10044", "10128")
#' zip_choropleth(df_pop_zip,
#'               zip_zoom = c(manhattan_les, manhattan_ues),
#'               title    = "2012 Lower and Upper East Side ZCTA Population Estimates",
#'               legend   = "Population")
#'
#' # zooming in on an entire Metropolitan Statistical Area (MSA)
#' zip_choropleth(df_pop_zip, 
#'                msa_zoom = "New York-Newark-Jersey City, NY-NJ-PA", 
#'                title    = "2012 NY-Newark-Jersey City MSA\nZCTA Population Estimates",
#'                legend   = "Population")
#'                
#'  # showing the entire country
#'  # note: this takes a few minutes to run
#'  zip_choropleth(df_pop_zip, 
#'                title  = "2012 US ZCTA Population Estimates", 
#'                legend = "Population")
#' }
#' @seealso \url{https://www.census.gov/geo/reference/zctas.html} for an explanation of ZCTAs and how they relate to US Zip Codes.
#' @export
#' @importFrom Hmisc cut2
#' @importFrom stringr str_extract_all
#' @importFrom ggplot2 ggplot aes geom_polygon scale_fill_brewer ggtitle theme theme_grey element_blank geom_text
#' @importFrom ggplot2 scale_fill_continuous scale_colour_brewer ggplotGrob annotation_custom 
#' @importFrom scales comma
zip_choropleth = function(df, title="", legend="", num_colors=7, state_zoom=NULL, county_zoom=NULL, msa_zoom=NULL, zip_zoom=NULL, reference_map=FALSE)
{
  # nationwide map is special - no borders and insets for AK and HI
  if (is.null(state_zoom) && is.null(county_zoom) && is.null(msa_zoom) && is.null(zip_zoom))
  {
    if (reference_map == TRUE)
    {
      stop("Reference maps do not currently work with maps that have insets, such as maps of the 50 US States.")
    }
    warning(paste0("Nationwide zip choropleths can take a few minutes to render. ",
            "It is much faster to view a subset of the country by selecting a zoom. ",
            "See ?zip_choropleth for zoom options."))
    
    c = ZipChoropleth$new(df)
    c$title  = title
    c$legend = legend
    c$set_num_colors(num_colors)
    c$render_nationwide()
  } else {
    c = ZipChoropleth$new(df)
    c$title  = title
    c$legend = legend
    c$set_zoom_zip(state_zoom=state_zoom, county_zoom=county_zoom, msa_zoom=msa_zoom, zip_zoom=zip_zoom)
    c$set_num_colors(num_colors)
    if (reference_map) {
      c$render_with_reference_map()
    } else {
      c$render()
    }
  }
}